lib/src/protobuf/coded_buffer_writer.dart

```diff
@@ -4,221 +4,494 @@
-    int lo = bytes.getUint32(0, Endian.little);
-    int hi = bytes.getUint32(4, Endian.little);
-    while (hi > 0 || lo >= 0x80) {
-      result[i++] = 0x80 | (lo & 0x7f);
-      lo = (lo >> 7) | ((hi & 0x7f) << 25);
-      hi >>= 7;
-    }
-    result[i++] = lo;
+  /// Number of bytes written into the [_outputChunk].
+  int _bytesInChunk = 0;
+
+  /// ByteData pointing to [_outputChunk]. Used to write primitive values
+  /// more efficiently.
+  ByteData _outputChunkAsByteData;
+
+  /// Array of pairs <Uint8List chunk, int bytesInChunk> - chunks are
+  /// pushed into this array once they are full.
+  final List<dynamic> _outputChunks = <dynamic>[];
 
-    return new ByteData.view(result.buffer, 0, i);
+  /// Total amount of bytes used in all chunks.
+  int _outputChunksBytes = 0;
+
+  /// Total amount of bytes written into this writer.
+  int _bytesTotal = 0;
+  int get lengthInBytes => _bytesTotal;
+
+  CodedBufferWriter() {
+    // Initialize [_outputChunk].
+    _commitChunk(true);
   }
 
-  static ByteData _int32ToBytes(int value) => _toVarint32(value & 0xffffffff);
+  void writeField(int fieldNumber, int fieldType, fieldValue) {
+    final int valueType = fieldType & ~0x07;
 
-  static _makeWriteFunctionMap() {
-    writeBytesNoTag(output, List<int> value) {
-      output.writeInt32NoTag(value.length);
-      output.writeRawBytes(
-          new Uint8List(value.length)..setRange(0, value.length, value));
+    if ((fieldType & PbFieldType._PACKED_BIT) != 0) {
+      if (!fieldValue.isEmpty) {
+        _writeTag(fieldNumber, WIRETYPE_LENGTH_DELIMITED);
+        final mark = _startLengthDelimited();
+        for (var value in fieldValue) {
+          _writeValueAs(valueType, value);
+        }
+        _endLengthDelimited(mark);
+      }
+      return;
     }
 
-    makeWriter(convertor) => ((output, value) {
-          output.writeRawBytes(convertor(value));
-        });
+    final int wireFormat = _wireTypes[_valueTypeIndex(valueType)];
 
-    int _encodeZigZag32(int value) => (value << 1) ^ (value >> 31);
+    if ((fieldType & PbFieldType._REPEATED_BIT) != 0) {
+      for (var i = 0; i < fieldValue.length; i++) {
+        _writeValue(fieldNumber, valueType, fieldValue[i], wireFormat);
+      }
+      return;
+    }
 
-    Int64 _encodeZigZag64(Int64 value) => (value << 1) ^ (value >> 63);
+    _writeValue(fieldNumber, valueType, fieldValue, wireFormat);
+  }
 
-    ByteData makeByteData32(int value) =>
-        new ByteData(4)..setUint32(0, value, Endian.little);
+  Uint8List toBuffer() {
+    Uint8List result = new Uint8List(_bytesTotal);
+    writeTo(result);
+    return result;
+  }
 
-    ByteData makeByteData64(Int64 value) {
-      var data = new Uint8List.fromList(value.toBytes());
-      return new ByteData.view(data.buffer, 0, 8);
+  /// Serializes everything written to this writer so far to [buffer], starting
+  /// from [offset] in [buffer]. Returns `true` on success.
+  bool writeTo(List<int> buffer, [int offset = 0]) {
```

Comment:
Does this need to be public?
That makes the writes to [buffer] potentially polymorphic.

Reply:
This method was added in a PR #70 by @postmasters, I don't know if this is really needed.